libraries/AP_Soaring/SoaringController.cpp

```diff
@@ -0,0 +1,391 @@
+    _prev_update_time = AP_HAL::micros64();
+    _thermal_start_time_us = AP_HAL::micros64();
+}
+   
+void SoaringController::init_cruising()
+{
+    _cruise_start_time_us = AP_HAL::micros64();
+    // Start glide. Will be updated on the next loop.
+    _throttle_suppressed = true;
+}
+
+void SoaringController::get_wind_corrected_drift(const Location *current_loc, const Vector3f *wind, float *wind_drift_x, float *wind_drift_y, float *dx, float *dy)
+{       
+    Vector2f diff = location_diff(_prev_update_location, *current_loc); // get distances from previous update
+    *dx = diff.x;
+    *dy = diff.y;
+
+    // Wind correction
+    *wind_drift_x = wind->x * (AP_HAL::micros64() - _prev_update_time) * 1e-6;
+    *wind_drift_y = wind->y * (AP_HAL::micros64() - _prev_update_time) * 1e-6;
+    *dx -= *wind_drift_x;
+    *dy -= *wind_drift_y;
+}
+
+void SoaringController::get_altitude_wrt_home(float *alt)
+{
+    _ahrs.get_relative_position_D_home(*alt);
+    *alt *= -1.0f;   
+}
+void SoaringController::update_thermalling()
+{
+    _loiter_rad = _aparm.loiter_radius;
+    struct Location current_loc;
+    _ahrs.get_position(current_loc);
+
+    if (_new_data) {  
+        float dx = 0;
+        float dy = 0;
+        float dx_w = 0;
+        float dy_w = 0;
+        Vector3f wind = _ahrs.wind_estimate();
+        get_wind_corrected_drift(&current_loc, &wind, &dx_w, &dy_w, &dx, &dy);
+
+        #if (0)
+            // Print32_t filter info for debugging
+            int32_t i;
+            for (i = 0; i < 4; i++) {
+                hal.console->printf("%e ", (double)_ekf.P[i][i]);
+            }
+            for (i = 0; i < 4; i++) {
+                hal.console->printf("%e ", (double)_ekf.X[i]);
+            }
+        #endif
+        
+        // write log - save the data.
+        log_tuning.time_us = AP_HAL::micros64();
+        log_tuning.netto_rate = _vario_reading;
+        log_tuning.dx = dx;
+        log_tuning.dy = dy;
+        log_tuning.x0 = _ekf.X[0];
+        log_tuning.x1 = _ekf.X[1];
+        log_tuning.x2 = _ekf.X[2];
+        log_tuning.x3 = _ekf.X[3];
+        log_tuning.lat = current_loc.lat;
+        log_tuning.lng = current_loc.lng;
+        log_tuning.alt = _alt;
+        log_tuning.dx_w = dx_w;
+        log_tuning.dy_w = dy_w;
+        log_tuning.nsamples = _nsamples;
+            
+        log_data(); 
+        _ekf.update(_vario_reading,dx, dy);       // update the filter
+         
+        _prev_update_location = current_loc;      // save for next time
+        _prev_update_time = AP_HAL::micros64();
+        _new_data = false;
+    }
+}
+
+void SoaringController::update_cruising()
+{
+    if (is_active() && suppress_throttle()) {
+        init_cruising();
+    }
+}
+
+void SoaringController::update_vario()
+{
+    Location current_loc;
+    _ahrs.get_position(current_loc);
+    //_alt = (current_loc.alt -_ahrs.get_home().alt) / 100.0f;
+    get_altitude_wrt_home(&_alt);
+    _alt *= -1.0f;    
```

Comment:
Are you sure about this? Your implementation of get_altitude_wrt_home() is already returning you a positive altitude, you just made it negative here, which I'm pretty sure is not your intention.